In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
from sklearn import datasets, linear_model
from sklearn.model_selection import train_test_split

In [2]:
#Import salary.csv as data using pd.read_csv method.
data = pd.read_csv('salary.csv')
data = pd.read_csv("salary.csv").dropna()

In [3]:
y = data.salary
X = data.drop('salary', axis=1)

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2)
print("X_train:")
print(X_train.head())
print(X_train.shape)

print("X_test:")
print(X_test.head())
print(X_test.shape)

X_train:
     exprior  yearsworked  yearsrank  market  degree  otherqual  position  \
101        1            8          3    1.21       1          0         2   
392        0            3          2    0.79       1          0         1   
182       15            3          3    0.92       1          0         3   
247        0           21          1    0.79       1          0         3   
217       13            8          8    1.05       1          0         1   

     male  Field  yearsabs  
101     1      1         1  
392     0      3         1  
182     1      3         0  
247     1      2         2  
217     1      3         0  
(410, 10)
X_test:
     exprior  yearsworked  yearsrank  market  degree  otherqual  position  \
325        0           26         20    0.93       1          0         2   
167        0           15          8    0.95       1          0         2   
259        0           25         15    1.14       1          0         3   
263        0           26   

The outcome variable is also called the response or dependent variable, and the risk factors and confounders are called the predictors, or explanatory or independent variables. In regression analysis, the dependent variable is denoted "Y" and the independent variables are denoted by "X". (http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_Correlation-Regression/BS704_Correlation-Regression_print.html). Yearsworked is the independant variable and salary is the dependant variable.

# 1. 

In [5]:
X = data['yearsworked'] #Predictor variable.
Y = data['salary']      #Response

In [6]:
X = sm.add_constant(X) # adding a constant term to the predictor


/home/nhlalwenhle/miniconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [7]:
X.head()

const  yearsworked
0    1.0            0
1    1.0            0
2    1.0            0
3    1.0            2
4    1.0            2

Ordinary Least Squares regression (OLS) is more commonly named linear regression (simple or multiple depending on the number of explanatory variables). smf stands for statsmodels formula.

In [8]:
model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 

In [9]:
print_model = model.summary()
print(print_model)

                            OLS Regression Results                            
Dep. Variable:                 salary   R-squared:                       0.389
Model:                            OLS   Adj. R-squared:                  0.388
Method:                 Least Squares   F-statistic:                     325.1
Date:                Wed, 15 May 2019   Prob (F-statistic):           1.29e-56
Time:                        11:14:10   Log-Likelihood:                -5448.0
No. Observations:                 513   AIC:                         1.090e+04
Df Residuals:                     511   BIC:                         1.091e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        4.012e+04    739.844     54.221      

Q1.1. Does the model significantly predict the dependent variable? Report the amount of variance explained (R^2) and significance value (p) to support your answer.

The model has a coefficent of determination of R^2 = 0.389 (38.9%). This means that the model has a regression line with 38.9% approximation of the real data points.
The p-value for both Const and Yearsworked are 0.000. A low p-value (< 0.05) indicates that you can reject the null hypothesis, meaning that a low p-value is likely to be a meaningful addition to your model because changes in the predictor's value are related to changes in the response variable. 
This means that the model can significantly predict the dependent variable (Salary).

Q1.2. What percentage of the variance in employees’ salaries is accounted for by the number of years they have worked?